packages/block-editor/src/hooks/border.js

```diff
@@ -39,22 +57,80 @@ export function BorderPanel( props ) {
 		return null;
 	}
 
+	const defaultBorderControls = getBlockSupport( props.name, [
+		BORDER_SUPPORT_KEY,
+		'__experimentalDefaultControls',
+	] );
+
+	const createResetAllFilter = (
+		borderAttribute,
+		topLevelAttributes = {}
+	) => ( newAttributes ) => ( {
+		...newAttributes,
+		...topLevelAttributes,
+		style: {
+			...newAttributes.style,
+			border: {
+				...newAttributes.style?.border,
+				[ borderAttribute ]: undefined,
+			},
+		},
+	} );
```

Comment:
Just a thought — should the reset all logic be applying/composing the individual `reset*` functions instead? (e.g. `resetBorderColor`, `resetBorderRadius`, `resetBorderStyle`, `resetBorderWidth`)

I see how it could be slightly less performant, but the new approach would avoid the duplication of the "resetting" logic in this file — WDYT?

Reply:
I'm not sure I understand what you are saying sorry. Are you asking whether we should be using the `resetAll` callback on the `ToolsPanel` itself here?

If so, we can't do that because these `ToolsPanelItem`s are being added to a `ToolsPanel` via a SlotFill. The slots provided by the grouped `InspectorControls` get wrapped in a `ToolsPanel`. That panel won't know what controls have been added inside the panel so each item needs to passing long what it needs to change in the attributes object when reset.

Reply:
Got it, I had overlooked the fact that the `ToolsPanelItem`s are being injected via SlotFill 😅 

I have another question then — could the `createResetAllFilter` function take a callback as an argument, instead of the `borderAttribute`? Something like `createResetAllFilter( resetBorderColor )` instead of `resetAllFilter={ createResetAllFilter( 'color', { borderColor: undefined, } ) }` ?

Excuses in advance in case this makes little sense.

Reply:
Maybe I'm being blind at the moment but I'm not seeing how the callback suggestion cleans up the current approach. 

We can't use `resetBorderColor` as that actually updates block attributes before all the other filters have been applied. In terms of a new callback, we'd be passing a callback to a higher-order function that's purpose is to create a callback. 

We could simplify `createResetAllFilter`by not having to deal with the top-level block attribute the border color needs to reset. It just means we need a custom function for the border color defined which makes things a touch more verbose.

<details>
  <summary>Here's a diff for what I mean</summary>
  
```diff
diff --git a/packages/block-editor/src/hooks/border.js b/packages/block-editor/src/hooks/border.js
index 83f0892b5c..c37dbeae03 100644
--- a/packages/block-editor/src/hooks/border.js
+++ b/packages/block-editor/src/hooks/border.js
@@ -62,12 +62,8 @@ export function BorderPanel( props ) {
 		'__experimentalDefaultControls',
 	] );
 
-	const createResetAllFilter = (
-		borderAttribute,
-		topLevelAttributes = {}
-	) => ( newAttributes ) => ( {
+	const createResetAllFilter = ( borderAttribute ) => ( newAttributes ) => ( {
 		...newAttributes,
-		...topLevelAttributes,
 		style: {
 			...newAttributes.style,
 			border: {
@@ -111,8 +107,16 @@ export function BorderPanel( props ) {
 					label={ __( 'Color' ) }
 					onDeselect={ () => resetBorderColor( props ) }
 					isShownByDefault={ defaultBorderControls?.color }
-					resetAllFilter={ createResetAllFilter( 'color', {
+					resetAllFilter={ ( newAttributes ) => ( {
+						...newAttributes,
 						borderColor: undefined,
+						style: {
+							...newAttributes.style,
+							border: {
+								...newAttributes.style?.border,
+								color: undefined,
+							},
+						},
 					} ) }
 					panelId={ clientId }
 				>
```
</details>


I've looked at this so many times I'm probably not seeing the forest for the trees. I don't feel strongly about the approach, so I'm happy to change this to whatever is needed to move this forward.

Reply:
Thank you for the explanation once again. I think the confusion here mostly come from me not fully grasping how the `resetAllFilter` is supposed to work.

> I don't feel strongly about the approach, so I'm happy to change this to whatever is needed to move this forward.

We can leave this as-is, and sorry if my review slowed this PR down more than it should have.